Add `async_fn_in_trait` lint

compiler-errors opened the pull request:
cc https://github.com/rust-lang/rust/pull/115822#issuecomment-1731168465

Mostly unsure what the messaging should be. Feedback required.

r? @tmandry

compiler-errors committed:
Add async_fn_in_trait lint

compiler-errors committed:
Bless tests

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6318129628/job/17156432038) [(plain)](https://github.com/rust-lang/rust/commit/b251f2cd2638ef268c08388847092c7a14178240/checks/17156432038/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:8ade135a41bc03ea155e62e844d188df1ea18608)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_4f41727d-5c6d-41f1-84e4-caa73dc19ea9
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=afit-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_4f41727d-5c6d-41f1-84e4-caa73dc19ea9
GITHUB_REF=refs/pull/116184/merge
GITHUB_REF_NAME=116184/merge
GITHUB_REF_PROTECTED=false
---
Removing intermediate container 3f21aa88a720
 ---> 1bc7b96a9330
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> a8e9bf6e908e
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-5l0m82yv/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.5-py3-none-any.whl (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 21.2 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 47.4 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.10.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.10.0 virtualenv-20.24.5
Removing intermediate container ec4695e43bd8
 ---> 3e0ef5528e27
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 472c21536686
 ---> 472c21536686
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 5643031ef926
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 62e1dfc25226
 ---> 28ee394860a8
Successfully built 28ee394860a8
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:28ee394860a8c3651046d073b1dd6574510a0e86fbd4f4d8afd8b8311de68121
Uploading finished image sha256:28ee394860a8c3651046d073b1dd6574510a0e86fbd4f4d8afd8b8311de68121 to https://ci-caches.rust-lang.org/docker/31024a61e3c099341c4151411b61536c5a15e77991c1c58c8adf475bd61d8568f58aad59ac3149cca672919d01d664a203adb87f86665b006743ece22accf3f1
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
28ee394860a8   1 second ago         /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
472c21536686   2 seconds ago        /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
3e0ef5528e27   3 seconds ago        |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
a8e9bf6e908e   10 seconds ago       /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
1bc7b96a9330   12 seconds ago       |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      5 weeks ago          /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      5 weeks ago          /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      5 weeks ago          /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fd4d0a2b690>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/31024a61e3c099341c4151411b61536c5a15e77991c1c58c8adf475bd61d8568f58aad59ac3149cca672919d01d664a203adb87f86665b006743ece22accf3f1
sha256:28ee394860a8c3651046d073b1dd6574510a0e86fbd4f4d8afd8b8311de68121
---
DirectMap4k:      169920 kB
DirectMap2M:     7170048 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6710e3312d54f03f6462481199b9e72367a2568b/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6710e3312d54f03f6462481199b9e72367a2568b-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint/src/async_fn_in_trait.rs:8: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint/src/async_fn_in_trait.rs:20: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint/src/async_fn_in_trait.rs:23: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint/src/async_fn_in_trait.rs:27: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.2.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 15.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 15.0 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 78.4 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:00:48
  local time: Tue Sep 26 20:51:05 UTC 2023
  network time: Tue, 26 Sep 2023 20:51:05 GMT
```

</details>

rakshith-ravi commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -5,6 +5,10 @@ lint_array_into_iter =
         or use `IntoIterator::into_iter(..)` instead of `.into_iter()` to explicitly iterate by value
 
+lint_async_fn_in_trait = usage of `async fn` in trait is discouraged because they do not automatically have auto trait bounds
+    .note = you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the future
```
Can we perhaps be more helpful to new users?  Like so:

```
you can suppress this lint if you do not care about auto traits like `Send` on the future with the `#[allow(async_fn_in_trait)]` attribute
```

compiler-errors replied:
I'd rather just make it a suggestion, but also, I don't think any other lint goes out of its way to tell you how to suppress it.

rakshith-ravi replied:
I'm sorry, but what's a structured suggestion?

Also, yes, other lints don't do this, but most other lints aren't elevated to be shown by default unless it's guaranteed to not have false positives. Additionally, new-comers probably wouldn't understand a lot about what `Send` bounds are and why it's needed. While yes, they should eventually learn it, it only increases the barrier to getting started with rust for them. Giving them a simple suggestions is more of a "here's the simple, ugly way out, but you should probably look into this when you understand what's happening"

rakshith-ravi replied:
On second thought, it would also encourage users to do the wrong thing unless they know what they're doing, which isn't exactly ideal.

Hmm, idk. I'm conflicted now. I've put both sides of the coin on the table. Open for suggestions

rakshith-ravi commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -5,6 +5,10 @@ lint_array_into_iter =
 
+lint_async_fn_in_trait = usage of `async fn` in trait is discouraged because they do not automatically have auto trait bounds
+    .note = you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the future
+    .suggestion = you can alternatively desugar the `async fn` and any add additional traits such as `Send` to the signature
```
```
you can alternatively desugar the `async fn` manually by making it a regular, non-async fn and returning an `impl std::future::Future<Output = *return-type*>` and any add additional traits such as `Send` to the signature
```

compiler-errors replied:
This is redundant given the contents of the suggestion showing you how to desugar immediately below this message

rakshith-ravi replied:
Ahh, okay. Got it.

rakshith-ravi reviewed (commented):
Just looking at this from the point of view of new comers, and I feel like explicit suggestions would be more helpful for them

traviscross committed:
Fill in prose to describe the `async_fn_in_trait` lint

We're stabilizing `async fn` in trait (AFIT), but we have some
reservations about how people might use this in the definitions of
publicly-visible traits, so we're going to lint about that.

This is a bit of an odd lint for `rustc`.  We normally don't lint just
to have people confirm that they understand how Rust works.  But in
this one exceptional case, this seems like the right thing to do as
compared to the other plausible alternatives.

In this commit, we describe the nature of this odd lint.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6364972293/job/17281636995) [(plain)](https://github.com/rust-lang/rust/commit/300123765e9feb5b029fb6e7b7ca38cbb6cbaa1e/checks/17281636995/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_c6efd2aa-b8fa-423a-9872-b84116d3a1a8
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=afit-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_c6efd2aa-b8fa-423a-9872-b84116d3a1a8
GITHUB_REF=refs/pull/116184/merge
GITHUB_REF_NAME=116184/merge
GITHUB_REF_PROTECTED=false
---

---- [ui] tests/ui/async-await/in-trait/warn.rs stdout ----
diff of stderr:

- error: usage of `async fn` in trait is discouraged because they do not automatically have auto trait bounds
+ error: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
3    |
4 LL |     async fn not_send();

5    |     ^^^^^
5    |     ^^^^^
6    |
-    = note: you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the future
+    = note: you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the `Future`
9   --> $DIR/warn.rs:4:9
10    |


11 LL | #![deny(async_fn_in_trait)]
12    |         ^^^^^^^^^^^^^^^^^
- help: you can alternatively desugar the `async fn` and any add additional traits such as `Send` to the signature
+ help: you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`
15 LL -     async fn not_send();
15 LL -     async fn not_send();
16 LL +     fn not_send() -> impl std::future::Future<Output = ()> + Send;

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/in-trait/warn/warn.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args async-await/in-trait/warn.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/async-await/in-trait/warn.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/in-trait/warn" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/in-trait/warn/auxiliary" "--edition=2021"
--- stderr -------------------------------
--- stderr -------------------------------
error: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
##[error]  --> /checkout/tests/ui/async-await/in-trait/warn.rs:7:5
LL |     async fn not_send();
Build completed unsuccessfully in 0:12:20
   |     ^^^^^
   |
   |
   = note: you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the `Future`
  --> /checkout/tests/ui/async-await/in-trait/warn.rs:4:9
   |
   |
LL | #![deny(async_fn_in_trait)]
   |         ^^^^^^^^^^^^^^^^^
help: you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`
LL -     async fn not_send();
LL +     fn not_send() -> impl std::future::Future<Output = ()> + Send;
   |

```

</details>

compiler-errors committed:
Only reachable traits

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6365137867/job/17281923796) [(plain)](https://github.com/rust-lang/rust/commit/353683e4e9590d80e82902e5d538dff15e0f4663/checks/17281923796/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=rust-cloud-vms[bot]
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a907f5a0-5270-462c-bcc2-3179006444b5
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=afit-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a907f5a0-5270-462c-bcc2-3179006444b5
GITHUB_REF=refs/pull/116184/merge
GITHUB_REF_NAME=116184/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=903b278e6bc3b314eda80cdc2566cf2ca3a16eee
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_a907f5a0-5270-462c-bcc2-3179006444b5
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_a907f5a0-5270-462c-bcc2-3179006444b5
GITHUB_TRIGGERING_ACTOR=rust-cloud-vms[bot]
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116184/merge
GITHUB_WORKFLOW_SHA=903b278e6bc3b314eda80cdc2566cf2ca3a16eee
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
##[endgroup]
 finished in 1.579 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
##[group]Running stage2 lint-docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `async_fn_in_trait` in /checkout/compiler/rustc_lint/src/async_fn_in_trait.rs:7: did not find lint `async_fn_in_trait` in output of example, got:


This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places
This tool extracts documentation for lints from the source code and places
them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html


To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
The --keep-stage flag should be used if you have already built the compiler

Build completed unsuccessfully in 0:24:29
  local time: Sat Sep 30 19:56:34 UTC 2023
  network time: Sat, 30 Sep 2023 19:56:34 GMT
```

</details>

compiler-errors commented on compiler/rustc_lint/src/async_fn_in_trait.rs:
````diff
@@ -0,0 +1,123 @@
+    /// For example, this code is invalid:
+    ///
+    /// ```rust,compile_fail
+    /// # #![feature(async_fn_in_trait)]
````
```suggestion
    /// # #![feature(async_fn_in_trait)]
    /// # #![deny(async_fn_in_trait)]
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6365757786/job/17283007638) [(plain)](https://github.com/rust-lang/rust/commit/c96e6502d1eca2bfdeb3ee3e9a83065b24a35f93/checks/17283007638/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_1b326ccb-614b-44d5-9964-3eabf2176ed6
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=afit-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_1b326ccb-614b-44d5-9964-3eabf2176ed6
GITHUB_REF=refs/pull/116184/merge
GITHUB_REF_NAME=116184/merge
GITHUB_REF_PROTECTED=false
---
##[endgroup]
 finished in 1.612 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
##[group]Running stage2 lint-docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `async_fn_in_trait` in /checkout/compiler/rustc_lint/src/async_fn_in_trait.rs:7: did not find lint `async_fn_in_trait` in output of example, got:


This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places
This tool extracts documentation for lints from the source code and places
them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html


To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
The --keep-stage flag should be used if you have already built the compiler

Build completed unsuccessfully in 0:24:47
  local time: Sat Sep 30 22:02:00 UTC 2023
  network time: Sat, 30 Sep 2023 22:02:00 GMT
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6382404213/job/17320931653) [(plain)](https://github.com/rust-lang/rust/commit/7e9496cf461a82195ba816a7e12c2cb1ce29a616/checks/17320931653/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=rust-cloud-vms[bot]
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_90d61629-2a9f-4e96-85dd-ccfd0b13b7b9
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=afit-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_90d61629-2a9f-4e96-85dd-ccfd0b13b7b9
GITHUB_REF=refs/pull/116184/merge
GITHUB_REF_NAME=116184/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=721f435632c93871880812b8641a9b20d5588254
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_90d61629-2a9f-4e96-85dd-ccfd0b13b7b9
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_90d61629-2a9f-4e96-85dd-ccfd0b13b7b9
GITHUB_TRIGGERING_ACTOR=rust-cloud-vms[bot]
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116184/merge
GITHUB_WORKFLOW_SHA=721f435632c93871880812b8641a9b20d5588254
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

failures:

---- compiler/rustc_lint/src/async_fn_in_trait.rs - async_fn_in_trait::ASYNC_FN_IN_TRAIT (line 12) stdout ----
error: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
  |
5 |     async fn method(&self);
  |     ^^^^^
  |
  |
  = note: you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the `Future`
 --> compiler/rustc_lint/src/async_fn_in_trait.rs:14:9
  |
  |
3 | #![deny(async_fn_in_trait)]
  |         ^^^^^^^^^^^^^^^^^
help: you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`
5 -     async fn method(&self);
5 -     async fn method(&self);
5 +     fn method(&self) -> impl std::future::Future<Output = ()> + Send;

error: aborting due to previous error

Couldn't compile the test.
Couldn't compile the test.

failures:
    compiler/rustc_lint/src/async_fn_in_trait.rs - async_fn_in_trait::ASYNC_FN_IN_TRAIT (line 12)

test result: FAILED. 80 passed; 1 failed; 2 ignored; 0 measured; 0 filtered out; finished in 2.69s

error: doctest failed, to rerun pass `-p rustc_lint --doc`
  local time: Mon Oct  2 16:10:25 UTC 2023
  network time: Mon, 02 Oct 2023 16:10:25 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tmandry commented on compiler/rustc_lint/src/async_fn_in_trait.rs:
```diff
@@ -0,0 +1,124 @@
+    /// }
+    ///
+    /// fn test<T: Trait>(x: T) {
+    ///     fn is_send<T: Send>(_: T) {}
+    ///     is_send(x.method()); // Not OK.
```
```suggestion
    ///     fn spawn<T: Send>(_: T) {}
    ///     spawn(x.method()); // Not OK.
```

tmandry commented on compiler/rustc_lint/src/async_fn_in_trait.rs:
````diff
@@ -0,0 +1,124 @@
+    /// }
+    /// ```
+    ///
+    /// Conversely, if the trait is used only locally, if only concrete types
+    /// that implement the trait are used, or if the trait author otherwise
+    /// does not care that the trait will not promise that the returned
+    /// [`Future`] implements any [auto traits] such as [`Send`], then the
+    /// lint may be suppressed.
````
```suggestion
    /// Conversely, if the trait is used only locally, if it is never used in
    /// generic functions, or if it is only used in single-threaded contexts
    /// that do not care whether the returned
    /// [`Future`] implements [auto traits] such as [`Send`], then the
    /// lint may be suppressed.
```

tmandry commented on tests/ui/async-await/in-trait/warn.stderr:
```diff
@@ -0,0 +1,20 @@
+LL |     async fn not_send();
+   |     ^^^^^
+   |
+   = note: you can suppress this lint if you plan to use the trait locally, for concrete types, or do not care about auto traits like `Send` on the `Future`
```
```suggestion
   = note: you can suppress this lint if you plan to use the trait only in your own code, or do not care about auto traits like `Send` on the `Future`
```

I think the emphasis on concrete types is confusing. I can't think of an example of a trait a user would write that is never used in generics.

tmandry commented on tests/ui/async-await/in-trait/warn.stderr:
```diff
@@ -0,0 +1,20 @@
+   |
+LL | #![deny(async_fn_in_trait)]
+   |         ^^^^^^^^^^^^^^^^^
+help: you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`
+   |
+LL -     async fn not_send();
+LL +     fn not_send() -> impl std::future::Future<Output = ()> + Send;
+   |
```
I think this should probably go first, above the note about suppressing.

compiler-errors replied:
This is not possible. By construction, we print all suggestions after notes:

https://github.com/rust-lang/rust/blob/2e5a9dd6c9eaa42f0684b4b760bd68fc27cbe51b/compiler/rustc_errors/src/emitter.rs#L2113-L2161

tmandry replied:
It's not the first time I've been annoyed by that, but oh well.

compiler-errors committed:
Address review nits

tmandry commented on compiler/rustc_lint/src/async_fn_in_trait.rs:
````diff
@@ -0,0 +1,123 @@
+    ///     fn method(&self) -> impl Future<Output = ()> + Send { async {} }
+    /// }
+    /// ```
+    ///
````
```suggestion
    ///
    /// This still allows the use of `async fn` within impls of the trait.
    /// However, it also means that the trait will never be compatible with
    /// implementations whose futures are not `Send`.
    ///
```

tmandry commented on compiler/rustc_lint/src/async_fn_in_trait.rs:
````diff
@@ -0,0 +1,123 @@
+    /// }
+    /// ```
+    ///
+    /// Conversely, if the trait is used only locally, if it is never used in
+    /// generic functions, or if it is only used in single-threaded contexts
+    /// that do not care whether the returned [`Future`] implements [auto traits]
+    /// such as [`Send`], then the lint may be suppressed.
````
Are we really concerned about other auto traits? I haven't run into a bound on futures that wasn't `Send`.

tmandry committed:
Apply suggestions from code review

Co-authored-by: Travis Cross <tc@traviscross.com>

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 2f5249019e94f99920dae6df832bde633b151eb4 has been approved by `tmandry`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tmandry 2f5249019e94f99920dae6df832bde633b151eb4 -->
<!-- homu: {"type":"Approved","sha":"2f5249019e94f99920dae6df832bde633b151eb4","approver":"tmandry","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2f5249019e94f99920dae6df832bde633b151eb4 with merge b781645332971d028d1d17c475fa2c40919132ba...
<!-- homu: {"type":"BuildStarted","head_sha":"2f5249019e94f99920dae6df832bde633b151eb4","merge_sha":"b781645332971d028d1d17c475fa2c40919132ba"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6413409812/job/17413924781)
Approved by: tmandry
Pushing b781645332971d028d1d17c475fa2c40919132ba to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"tmandry","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6413409812/job/17413924781"},"merge_sha":"b781645332971d028d1d17c475fa2c40919132ba"} -->

rust-timer commented:
Finished benchmarking commit (b781645332971d028d1d17c475fa2c40919132ba): [comparison URL](https://perf.rust-lang.org/compare.html?start=afe67fa2ef1bb0dcf9077d55cdd1c64a9969c9cf&end=b781645332971d028d1d17c475fa2c40919132ba&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=afe67fa2ef1bb0dcf9077d55cdd1c64a9969c9cf&end=b781645332971d028d1d17c475fa2c40919132ba&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=afe67fa2ef1bb0dcf9077d55cdd1c64a9969c9cf&end=b781645332971d028d1d17c475fa2c40919132ba&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 4.0% | [3.0%, 5.0%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=afe67fa2ef1bb0dcf9077d55cdd1c64a9969c9cf&end=b781645332971d028d1d17c475fa2c40919132ba&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=afe67fa2ef1bb0dcf9077d55cdd1c64a9969c9cf&end=b781645332971d028d1d17c475fa2c40919132ba&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 622.444s -> 623.73s (0.21%)
**Artifact size**: 272.02 MiB -> 272.01 MiB (-0.01%)

jonhoo commented:
The backwards compatibility hazard around going the `+ Send` route isn't currently highlighted by the lint. I think it's essential to add that in there, because as written folks will see the suggestion and go "oh, sweet, that fixes my problem, let me just do that!", not realizing that they've now committed to `+ Send` until the next major release.

compiler-errors commented:
I'll make a TODO to make that present in the diagnostic. For the record, the lint description does note this:
```
    /// This still allows the use of `async fn` within impls of the trait.
    /// However, it also means that the trait will never be compatible with
    /// impls where the returned [`Future`] of the method does not implement
    /// `Send`.
```

jonhoo commented:
Yeah, I think it specifically needs to be mentioned in the `.suggestion` — I've found that it's pretty rare someone actually reads (or even knows how to read) lint descriptions.

I saw that bit, and it's close, but I think missing a crucial bi-sentence. Something like:

> ... without a breaking change to the trait where the `+ Send` bound is removed again.

tmandry commented:
> Yeah, I think it specifically needs to be mentioned in the `.suggestion` — I've found that it's pretty rare someone actually reads (or even knows how to read) lint descriptions.

I don't agree that it should go in the suggestion; the implication of the syntax should really be clear to anyone who is publishing a trait for public consumption. If someone is unsure, that's a good reason to dig deeper by reading the lint description, the blog post we'll publish about releasing AFIT, etc.

> ... without a breaking change to the trait where the `+ Send` bound is removed again.

I'm not opposed to saying something like this, though editorially, it's already a pretty long sentence. I'm happy to review a PR.

jonhoo commented:
> the implication of the syntax should really be clear to anyone who is publishing a trait for public consumption

I agree on the "should", but disagree on the implied "would". My hope was that we could help trait definers understand the trade-off they're making, but if you feel strongly that the backwards-compat hazard here will be obvious to all relevant users, then I won't belabor the point. I just don't share your conviction that that is true :sweat_smile: 

> dig deeper by reading the lint description

I didn't even known lints _had_ descriptions until this discussion. How does one find them? I can't find a reference anywhere to how to print a lint description? Does it require digging into the rustc src? If so, I think no-one who's at risk from this hazard will go dig that up.

> the blog post we'll publish about releasing AFIT

Here's a thought: maybe we could link to the blog post from the lint (once it's out, so one release later)? That feels like it might be a good way to give folks more context than the little note.

tmandry commented:
> How does one find them?

Normally the error comes with instructions to run a command like `rustc --explain E0046` to print them, but I just noticed this lint doesn't have an error code. @compiler-errors how can we make this description accessible?

For what it's worth, they're also available online in the [error codes index](https://doc.rust-lang.org/error_codes/error-index.html). I don't expect anyone to be digging into the compiler source :).

> maybe we could link to the blog post from the lint

We could link to it from the lint or from the lint description, yeah.

compiler-errors commented:
@tmandry: As far as I can tell, lints are not typically associated with error codes. The only one I could find is E0602, but that doesn't use the same hir-based lint infrastructure that we do to emit this lint. Error code docs would also need to copy the lint description, since they're populated from separate markdown files.

If we think additional information should be attached, then instead of error code docs, I think we should just put another note on the diagnostic.

jonhoo commented:
According to [the rustc dev guide](https://rustc-dev-guide.rust-lang.org/diagnostics.html#diagnostic-structure) at least:

> Diagnostics created by lints don't have a code in the emitted message.

Which matches what I thought was the case :sweat_smile: And I agree with @compiler-errors that, given this is the case, we should probably surface some more info from the description in the notes.
